From ce5778da4656a89ecb630e1719fccbf85f44d29d Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 11 Jul 2025 19:13:34 +0200
Subject: [PATCH 896/954] plat: stm32mp2: add stm32mp_ca35_configure_opp

Add a new function stm32mp_ca35_configure_opp() used to manage OPP change.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1e7f4c816cbb621aa415e0f419c3150a8566db7d plat: stm32mp2: add stm32mp_ca35_configure_opp]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I9c15f43d61b240dc84bbe77e8c98a6b33b44c0a1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/475533
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/include/stm32mp2_private.h |  1 +
 plat/st/stm32mp2/stm32mp2_ca35ss.c          | 72 +++++++++++++++++++++
 2 files changed, 73 insertions(+)

diff --git a/plat/st/stm32mp2/include/stm32mp2_private.h b/plat/st/stm32mp2/include/stm32mp2_private.h
index 816a1ae320..3ac098842b 100644
--- a/plat/st/stm32mp2/include/stm32mp2_private.h
+++ b/plat/st/stm32mp2/include/stm32mp2_private.h
@@ -118,5 +118,6 @@ uint32_t otp_mirror_read(uint32_t *val, uint32_t otp);
 void stm32mp_ca35_set_vbar(uintptr_t vbar);
 void stm32mp_ca35_lpi_isolate(void);
 void stm32mp_ca35_lpi_restore(void);
+void stm32mp_ca35_configure_opp(bool overdrive);
 
 #endif /* STM32MP2_PRIVATE_H */
diff --git a/plat/st/stm32mp2/stm32mp2_ca35ss.c b/plat/st/stm32mp2/stm32mp2_ca35ss.c
index 521f34ebf5..2a13248e08 100644
--- a/plat/st/stm32mp2/stm32mp2_ca35ss.c
+++ b/plat/st/stm32mp2/stm32mp2_ca35ss.c
@@ -24,6 +24,8 @@
  */
 #define CA35SS_SSC_LPI_TSGEN_NTS(type)	(U(0x0D0) + A35SSC_SSC_ ## type)
 #define CA35SS_SSC_LPI_STGEN_NTS(type)	(U(0x140) + A35SSC_SSC_ ## type)
+#define CA35SS_SSC_OPP_REQ(type)	(U(0x1A0) + A35SSC_SSC_ ## type)
+#define CA35SS_SSC_MEM_CTRL(type)	(U(0x1C0) + A35SSC_SSC_ ## type)
 
 #define CA35SS_SYSCFG_VBAR_CR		0x2084U
 
@@ -38,6 +40,20 @@
  */
 #define CA35SS_SSC_LPI_STGEN_CSYSREQ			BIT_32(24)
 #define CA35SS_SSC_LPI_STGEN_CSYSACK			BIT_32(25)
+
+/*
+ * CA35SS_SSC_OPP_REQ register fields
+ */
+#define CA35SS_SSC_OPP_REQ_REQ				BIT_32(0)
+#define CA35SS_SSC_OPP_REQ_ACK				BIT_32(1)
+
+/*
+ * CA35SS_SSC_MEM_CTRL register fields
+ */
+#define CA35SS_SSC_MEM_CTRL_RM_MASK			GENMASK_32(2, 0)
+#define CA35SS_SSC_MEM_CTRL_RM_OVERDRIVE		U(3)
+#define CA35SS_SSC_MEM_CTRL_RME				BIT_32(3)
+
 #define TIMEOUT_US			U(1000)
 
 void stm32mp_ca35_set_vbar(uintptr_t vbar)
@@ -113,3 +129,59 @@ void stm32mp_ca35_lpi_restore(void)
 		}
 	}
 }
+
+/*
+ * Configure CA35SS for new OPP with syscfg register SSC_MEM and OPP_REQ
+ * the parameter overdrive indicates if the new OPP has an overdrive frequency
+ * NB: as STGEN is deactivated during OPP request, use polling without timeout
+ */
+void stm32mp_ca35_configure_opp(bool overdrive)
+{
+	uint32_t mem_ctrl = mmio_read_32(A35SSC_BASE + CA35SS_SSC_MEM_CTRL(RW));
+	uint32_t counter = UINT32_MAX;
+
+	/*
+	 * SSC_MEM control the speed of memories inside Cortex-A35
+	 * - RME: 0 default margin setting for nominal Cortex-A35 frequencies
+	 * - RME: 1 margin setting from RM[2:0]
+	 * - RM: CA35SS_SSC_MEM_CTRL_RM_OVERDRIVE the recommended value for
+	 *       Cortex-A35 overdrive frequencies, not used if RME = 0
+	 *
+	 * This function configure the CA35SS values only if the OPP change
+	 * (nominal vs overdrive)
+	 */
+	if (overdrive) {
+		if ((mem_ctrl & CA35SS_SSC_MEM_CTRL_RME) == CA35SS_SSC_MEM_CTRL_RME) {
+			return;
+		}
+		mmio_write_32(A35SSC_BASE + CA35SS_SSC_MEM_CTRL(WC1),
+			      CA35SS_SSC_MEM_CTRL_RM_MASK);
+		mmio_write_32(A35SSC_BASE + CA35SS_SSC_MEM_CTRL(WS1),
+			      CA35SS_SSC_MEM_CTRL_RME | CA35SS_SSC_MEM_CTRL_RM_OVERDRIVE);
+	} else {
+		if ((mem_ctrl & CA35SS_SSC_MEM_CTRL_RME) == 0U) {
+			return;
+		}
+		mmio_write_32(A35SSC_BASE + CA35SS_SSC_MEM_CTRL(WC1), CA35SS_SSC_MEM_CTRL_RME);
+	}
+
+	stm32mp_ca35_lpi_isolate();
+
+	/* Launch the OPP request*/
+	mmio_write_32(A35SSC_BASE + CA35SS_SSC_OPP_REQ(WS1), CA35SS_SSC_OPP_REQ_REQ);
+
+	/* Wait for the request to be complete */
+	while ((mmio_read_32(A35SSC_BASE + CA35SS_SSC_OPP_REQ(RW))
+		& CA35SS_SSC_OPP_REQ_ACK) == 0U) {
+		/* After LPI isolate, the ARM generic timer is frozen  */
+		counter--;
+		if (counter == 0U) {
+			panic();
+		}
+	}
+
+	/* Stop the OPP request */
+	mmio_write_32(A35SSC_BASE + CA35SS_SSC_OPP_REQ(WC1), CA35SS_SSC_OPP_REQ_REQ);
+
+	stm32mp_ca35_lpi_restore();
+}
-- 
2.53.0

